ttnn/cpp/ttnn/operations/moreh/moreh_linear_backward/device/moreh_linear_backward_device_operation.cpp

```diff
@@ -0,0 +1,115 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_linear_backward_device_operation.hpp"
+
+#include <cstdint>
+
+#include "tt_dnn/op_library/moreh_helper_functions.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::moreh::moreh_linear_backward {
+
+void MorehBiasAddBackwardOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto& output_grad = tensor_args.output_grad;
+    auto& input = tensor_args.input;
+    auto& weight = tensor_args.weight;
+    auto& input_grad = tensor_args.input_grad;
+    auto& weight_grad = tensor_args.weight_grad;
+    auto& bias_grad = tensor_args.bias_grad;
+
+    if (input_grad.has_value()) {
+        auto input_shape = input.get_shape();
+        auto input_grad_shape = input_grad->get_shape();
+        TT_ASSERT(input_shape == input_grad_shape, "both tensors should be the same shape");
+    }
+
+    if (weight_grad.has_value()) {
+        auto weight_shape = weight.get_shape();
+        auto weight_grad_shape = weight_grad->get_shape();
+        TT_ASSERT(weight_shape == weight_grad_shape, "both tensors should be the same shape");
+    }
+
+    if (bias_grad.has_value()) {
+        auto bias_grad_shape = bias_grad->get_shape();
+        auto bias_grad_tensor = bias_grad.value();
+        TT_ASSERT(
+            tt::operations::primary::is_scalar(bias_grad_tensor) ||
+                tt::operations::primary::is_1d_tensor(bias_grad_tensor),
+            "bias_grad tensor should be 1d or scalar");
+    }
+}
+
+MorehBiasAddBackwardOperation::program_factory_t MorehBiasAddBackwardOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& bias_grad = tensor_args.bias_grad.value();
+    if (tt::operations::primary::is_scalar(bias_grad))
+        return SingleCoreProgramFactory();
+    return MultiCoreProgramFactory();
```

Comment:
If a different factory is called based on bias, it should have a name other than SingleCore factory.

Reply:
`operation::ProgramWithCallbacks MorehBiasAddBackward::create_program(
    const std::vector<Tensor>& inputs, std::vector<Tensor>& outputs) const {
    const auto& output_grad = inputs.at(0);
    const auto& bias_grad = outputs.at(0);
    return is_scalar(bias_grad) ? (moreh_bias_backward_single_core_hw(output_grad, bias_grad, this->compute_kernel_config))
                                : (moreh_bias_backward_multi_core_h(output_grad, bias_grad, this->compute_kernel_config));
}`

This is the old implementation code so I just copy the program factory name, do you think I should change it then?

Reply:
Oh, then it seems fine to keep it as is for now.